### DISCUSSION TITLE
Performance tests (benchmarks) is required

### DISCUSSION THREAD

**Discussion by bekliev** [alpinejs/alpine#632] [4. General]
A while ago I've created a couple of PRs:
1. https://github.com/alpinejs/alpine/pull/629 (mutation observer)
2. https://github.com/alpinejs/alpine/pull/630 (transitions)
3. https://github.com/alpinejs/alpine/pull/631 (getXAttrs)

But there's a problem with determining of usefulness of these PRs, cause at the moment we don't have ability to compare performance BEFORE and AFTER changes in those PRs.

I've made some isolated tests for the approach I've implemented in those PRs, but it doesn't show clearly how it will affect in real application in runtime.

## I suggest to create benchmarks specifically for this library

I've never done benchmarks for libraries but only small performance tests.
And I even cannot imagine how the benchmark for this library should look like.
How many cases should be covered and etc.

Let's find out together.

Any thoughts?

**Comment by ryangjchandler** [comment]
You have to think about what is important for the library in terms of performance.

For most JavaScript libraries, that means how long it takes to completely initialise and how performant DOM operations are.

The first would involve both Alpine running the DOM walks as well as any initial DOM operations for bindings, visibility and actual node presence.

You could even micro-benchmark / unit-benchmark things like the `saferEval` and utility functions since they are the backbone of Alpine.

**Reply by bekliev** [reply]
I meant runtime performance.
In PR https://github.com/alpinejs/alpine/pull/629 there is "improvement" (which has to be proven) which makes sense only for runtime DOM mutations after Alpine finished its first parsings/bindings/executions.

For example: user clicks on button which opens modal window which contents is generated by server (like Ruby on Rails, PHP, etc.).
And one of the DOM elements contain `x-data` attribute which will be processed by Mutation Observer.

So, to determin which way is faster in this case it's better to have some kind of benchmark which will create some DOM-elements with `x-data` attribute to trigger Mutation Observer.

Then we can call it "Mutation Observer Performance Test" which is going to be one of the many tests out there.
